### DISCUSSION TITLE
Use Golang native features for building dependencies

### DISCUSSION THREAD

**Discussion by Sharpz7** [armadaproject/armada#1846] [Ideas]
A wider discussion based on https://github.com/G-Research/armada/issues/1836

Essentially, we should look into using more golang native technoloigies for building the tools we need for running docker in production and in a developer enviroment. I think this would also be an easier, more modular approach to starting the work on our operator idea (https://github.com/G-Research/armada/discussions/1796).

**Comment by Sharpz7** [comment]
> (Inspired by @jgiannuzzi and https://github.com/hashicorp/vault/tree/main/helper/testhelpers)
> 
> Essentially, instead of using the makefile, the proto-file generation, Docker/Kind Setup and in theory any files that need to be created, can all by handled from go instead of the makefle using `go build` - creating binary "plugins" with the same behavior.
> 
> I think in general, vault seems to have followed all these "go" good practises, so I think constant referal to their Makefile would be useful (https://github.com/hashicorp/vault/blob/main/Makefile).

Taken from the issue - essentially it's using golang instead of bash to write our setup code.

**Comment by suprjinx** [comment]
I'm not sure I see a huge advantage here -- the Vault project still has a giant makefile full of bash commands

**Reply by Sharpz7** [reply]
I would say that their makefile is cleaner - codewise, as well as having golang code that is better suited to catch errors - that's the real advantage I see
